### PR TITLE
Fix all warnings. Treat new ones as errors

### PR DESCRIPTION
**Pull request by ben-clayton**

### CODE REVIEW

File: core/cc/id.h [hunk 1]
```diff
@@ -51,7 +51,12 @@ namespace std {
 
 template <> struct hash<core::Id> {
     inline size_t operator()(const core::Id& id) const {
-        return *reinterpret_cast<const size_t*>(id.data);
+        union {
```
Comment (AWoloszyn) [inline core/cc/id.h]: IIRC this breaks strict aliasing.
The only way to do this kind of cast "legally" is
```
static_assert(sizeof(core::Id*) == sizeof(size_t), "Incompatible types"); // Sanity check
size_t hash;
memcpy(&hash, &id, sizeof(size_t));
return hash;

All of the compilers SHOULD optimize the memcpy away, and this becomes a no-op anyway.

File: gapil/runtime/cc/string_test.cpp [hunk 2]
```diff
@@ -44,7 +44,7 @@ TEST(StringTest, constructors) {
         EXPECT_STREQ(strB.c_str(), abc123);
     }
     {
-        gapil::String str(std::move(gapil::String(&arena, abc123)));
+        gapil::String str(gapil::String(&arena, abc123));
```
Comment (AWoloszyn) [inline gapil/runtime/cc/string_test.cpp]: Can this not be
```
gapil::String str(&arena, abc123);
```

File: core/cc/id.h (hunk 1 above)
Comment (ben-clayton) [reply core/cc/id.h]: Done. My memory is set on telling me that the `union` is the correct solution, regardless on how many times you correct me. :)

File: gapil/runtime/cc/string_test.cpp (hunk 2 above)
Comment (AWoloszyn) [reply gapil/runtime/cc/string_test.cpp]: Sorry my bad!

File: core/cc/id.h (hunk 1 above)
Comment (AWoloszyn) [reply core/cc/id.h]: Looks good!

**Comment by ben-clayton** [comment]
Thanks. I'll be appending minor fixes until the build machines are green.

**Comment by ben-clayton** [comment]
The windows build is having unrelated issues which I'm already looking into.
Submitting.

**Pull request merged**